iceoryx_posh/source/roudi/process_manager.cpp

```diff
@@ -0,0 +1,685 @@
+            << "Version mismatch from '" << name
+            << "'! Please build your app and RouDi against the same iceoryx version (version & commitID). RouDi: "
+            << version::VersionInfo::getCurrentVersion().operator iox::cxx::Serialization().toString()
+            << " App: " << versionInfo.operator iox::cxx::Serialization().toString();
+        return false;
+    }
+    // overflow check
+    if (m_processList.size() >= MAX_PROCESS_NUMBER)
+    {
+        LogError() << "Could not register process '" << name << "' - too many processes";
+        return false;
+    }
+    m_processList.emplace_back(name, pid, *payloadMemoryManager, isMonitored, payloadSegmentId, sessionId);
+
+    // send REG_ACK and BaseAddrString
+    runtime::IpcMessage sendBuffer;
+
+    auto offset = RelativePointer::getOffset(m_mgmtSegmentId, m_segmentManager);
+    sendBuffer << runtime::IpcMessageTypeToString(runtime::IpcMessageType::REG_ACK)
+               << m_roudiMemoryInterface.mgmtMemoryProvider()->size() << offset << transmissionTimestamp
+               << m_mgmtSegmentId;
+
+    m_processList.back().sendViaIpcChannel(sendBuffer);
+
+    // set current timestamp again (already done in Process's constructor
+    m_processList.back().setTimestamp(mepoo::BaseClock_t::now());
+
+    m_processIntrospection->addProcess(static_cast<int>(pid), ProcessName_t(cxx::TruncateToCapacity, name.c_str()));
+
+    LogDebug() << "Registered new application " << name;
+    return true;
+}
+
+bool ProcessManager::unregisterProcess(const ProcessName_t& name) noexcept
+{
+    if (!searchForProcessAndRemoveIt(name))
+    {
+        LogError() << "Application " << name << " could not be unregistered!";
+        return false;
+    }
+    return true;
+}
+
+bool ProcessManager::searchForProcessAndRemoveIt(const ProcessName_t& name) noexcept
+{
+    // we need to search for the process (currently linear search)
+    auto it = m_processList.begin();
+    while (it != m_processList.end())
+    {
+        auto otherName = it->getName();
+        if (name == otherName)
+        {
+            if (removeProcessAndDeleteRespectiveSharedMemoryObjects(it))
+            {
+                LogDebug() << "Removed existing application " << name;
+            }
+            return true; // we can assume there are no other processes with this name
+        }
+        ++it;
+    }
+    return false;
+}
+
+bool ProcessManager::removeProcessAndDeleteRespectiveSharedMemoryObjects(ProcessList_t::iterator& processIter) noexcept
+{
+    if (processIter != m_processList.end())
+    {
+        m_portManager.deletePortsOfProcess(processIter->getName());
+        m_processIntrospection->removeProcess(static_cast<int32_t>(processIter->getPid()));
+
+        // Reply with TERMINATION_ACK and let process shutdown
+        runtime::IpcMessage sendBuffer;
+        sendBuffer << runtime::IpcMessageTypeToString(runtime::IpcMessageType::TERMINATION_ACK);
+        processIter->sendViaIpcChannel(sendBuffer);
```

Comment:
this should not be done here but in roudi.cpp where also the `runtime::IpcMessageType::TERMINATION` is handled. This method is also called without being invoked by an application

Reply:
Good catch! Unfortunately, at the level of roudi.cpp there is no way of sending a `IpcMessage` to a specific process as the iterator is needed.

I would add a 2nd parameter to `searchForProcessAndRemoveIt(name, DoNotSendAck)` to be able to omit the `TERMINATION_ACK` message. What do you think?